### PR TITLE
DOC: Tutorial for least_squares

### PR DESCRIPTION
**Pull request by nmayorov**
Hi!

This is a promised update of the tutorial on least-squares minimization. The section itself turned out to be very small, the idea is to compensate for that by 3 examples. Currently I give links to gists on my github account. I hope we can come up with a better solution, https://scipy-cookbook.readthedocs.org from @pv looks pretty cool, maybe we can put it there already?

### CODE REVIEW

**Comment by pv** [comment]
Probably doesn't do harm to put them also there, even if we decide later on a different place. You can send a pull request with the notebook there (maybe add line "TAGS: Numpy / Scipy; Optimization and fitting techniques" to the notebook text).

**Comment by nmayorov** [comment]
Now example links lead to http://scipy-cookbook.readthedocs.org/

Waiting for the feedback about the content.

**Comment by nmayorov** [comment]
So can it go into the docs? 

@ev-br @pv any feedback?

**Comment by ev-br** [comment]
Test failures are unrelated, and can be ignored. Example notebooks look super cool. 
One problem is that all three examples have quite a few bells and whistles, so that they might be less than accessible to novices. To this end, I think we should have a small nearly trivial example showing how to set up the simplest possible problem for `least_squares`, and then link to cool examples.

**Comment by nmayorov** [comment]
I included a rather straightforward (yet meaningful) example. Indeed now this section in tutorial looks more solid. 

Can you do another look through?

File: doc/source/tutorial/optimize.rst [hunk 1]
```diff
@@ -368,106 +368,129 @@ and a constrained optimization as:
...
 .. math::
-   :nowrap:
-
-    e_{i}\left(\mathbf{p},\mathbf{y}_{i},\mathbf{x}_{i}\right)=\left\Vert \mathbf{y}_{i}-\mathbf{f}\left(\mathbf{x}_{i},\mathbf{p}\right)\right\Vert
-
-An objective function to pass to any of the previous minimization
-algorithms to obtain a least-squares fit is.
+  \begin{align}
+  &\min_\mathbf{x} \frac{1}{2} \sum_{i = 1}^m \rho\left(f_i(\mathrm{x})^2\right) \\
+  &\text{s. t. } \mathbf{lb} \leq \mathbf{x} \leq \mathbf{ub}
```
Comment (rgommers) [inline doc/source/tutorial/optimize.rst]: I haven't actually seen ``s.t.`` before. what does it stand for?

File: doc/source/tutorial/optimize.rst [hunk 2]
```diff
@@ -368,106 +368,129 @@ and a constrained optimization as:
...
+  \end{align}
+
+Here :math:`f_i(\mathbf{x})` are smooth functions from
+:math:`\mathbb{R}^n` to :math:`\mathbb{R}`, we refer to them as residuals.
+The purpose of a scalar valued function :math:`\rho(\cdot)` is to reduce
+influence of outlier residuals and contribute to robustness of the solution,
```
Comment (rgommers) [inline doc/source/tutorial/optimize.rst]: nitpick: "the influence"

File: doc/source/tutorial/optimize.rst [hunk 3]
```diff
@@ -368,106 +368,129 @@ and a constrained optimization as:
...
-   :nowrap:
+    \begin{align}
+    &J_{i0} = \frac{\partial f_i}{\partial x_0} = \frac{u_i^2 + u_i x_1}{u_i^2 + u_i x_2 + x_3} \\
+    &J_{i1} = \frac{\partial f_i}{\partial x_1} = \frac{u_i x_0}{u_i^2 + u_i x_2 + x_3} \\
+    &J_{i2} = \frac{\partial f_i}{\partial x_2} = -\frac{x_0 (u_i^2 + u_i x_1) u_i}{(u_i^2 + u_i x_2 + x_3)^2} \\
+    &J_{i3} = \frac{\partial f_i}{\partial x_3} = -\frac{x_0 (u_i^2 + u_i x_1)}{(u_i^2 + u_i x_2 + x_3)^2}
+    \end{align}
 
-    y_{i}=A\sin\left(2\pi kx_{i}+\theta\right)
+We are going to use "hard" starting point defined in [1]_. To find a physically
```
Comment (rgommers) [inline doc/source/tutorial/optimize.rst]: nitpick: ``use the "hard"``

File: doc/source/tutorial/optimize.rst (hunk 1 above)
Comment (nmayorov) [inline doc/source/tutorial/optimize.rst]: I thought of "subject to". I guess it is not really a valid acronym, not sure where I saw it used.

Comment (rgommers) [inline doc/source/tutorial/optimize.rst]: maybe just remove it? don't think you need anything there,

**Comment by rgommers** [comment]
I've read through the html output, all looks good and the example is simple to understand. +1 from me

File: doc/source/tutorial/optimize.rst (hunk 1 above)
Comment (ev-br) [inline doc/source/tutorial/optimize.rst]: I think I came across s.t. standing for "such that". As in "For every epsilon larger than zero there exists delta larger than then zero such that ...".

(No opinion on whether this particular case is better with or without it)

Comment (charris) [inline doc/source/tutorial/optimize.rst]: Mathematicians use "s.t." for "such that".

**Comment by nmayorov** [comment]
Ralf's notes were fixed. 

I changed to "subject to", it is consistent with other sections in the optimize tutorial.

**Comment by rgommers** [comment]
Thanks for the clarifications all. +1 for merging this once TravisCI is happy.

**Comment by ev-br** [comment]
Test run is all green, so I hit the green button. Thanks Nikolay, all

**Comment by anntzer** [comment]
A few ideas:
1. Include some discussion about using `diag(J^T J)` as an approximation of the covariance (and how this is affected by weighting).
2. Change the colormap of the plots in "Solving a discrete boundary-value problem in scipy" so that (a) white corresponds to zero in both cases (in the first solution this can be done via `vmin=-max(abs(res.x)), vmax=max(abs(res.x))`; in the second solution I would set the colormap to `"Reds"`).

**Comment by nmayorov** [comment]
@anntzer 

2. Great advice, looks much cleaner now! I sent a PR https://github.com/pv/SciPy-CookBook/pull/9

1. I didn't pay much attention to this aspect (it didn't look very important to me). But I noticed that a lot of people are interested about this covariance estimation, so I'll try to include some info.

**Comment by anntzer** [comment]
Actually I just noticed that `curve_fit` also gained the ability to call `least_squares` internally, that is great (and should probably be emphasized)!  I also noticed that there's also some code for computing the covariance there, so even a pointer to that would be good.
Covariance is useful as a first approximation of confidence intervals (MCMC is probably the proper way to dot it but it's quite a bit more expensive).

**Pull request merged**